source/meshworker/scratch_data.cc

```diff
@@ -0,0 +1,356 @@
+    const typename DoFHandler<dim, spacedim>::active_cell_iterator &cell,
+    const unsigned int                                              face_no)
+  {
+    if (!fe_face_values)
+      fe_face_values = std::make_unique<FEFaceValues<dim, spacedim>>(
+        *mapping, *fe, face_quad, face_flags);
+
+    fe_face_values->reinit(cell, face_no);
+    cell->get_dof_indices(local_dof_indices);
+    current_fe_values = fe_face_values.get();
+    return *fe_face_values;
+  }
+
+
+
+  template <int dim, int spacedim>
+  const FEFaceValuesBase<dim, spacedim> &
+  ScratchData<dim, spacedim>::reinit(
+    const typename DoFHandler<dim, spacedim>::active_cell_iterator &cell,
+    const unsigned int                                              face_no,
+    const unsigned int                                              subface_no)
+  {
+    if (subface_no != numbers::invalid_unsigned_int)
+      {
+        if (!fe_subface_values)
+          fe_subface_values = std::make_unique<FESubfaceValues<dim, spacedim>>(
+            *mapping, *fe, face_quad, face_flags);
+        fe_subface_values->reinit(cell, face_no, subface_no);
+        cell->get_dof_indices(local_dof_indices);
+
+        current_fe_values = fe_subface_values.get();
+        return *fe_subface_values;
+      }
+    else
+      return reinit(cell, face_no);
+  }
+
+
+
+  template <int dim, int spacedim>
+  const FEValues<dim, spacedim> &
+  ScratchData<dim, spacedim>::reinit_neighbor(
+    const typename DoFHandler<dim, spacedim>::active_cell_iterator &cell)
+  {
+    if (!neighbor_fe_values)
+      neighbor_fe_values = std::make_unique<FEValues<dim, spacedim>>(
+        *mapping, *fe, cell_quad, neighbor_cell_flags);
+
+    neighbor_fe_values->reinit(cell);
+    cell->get_dof_indices(neighbor_dof_indices);
+    current_neighbor_fe_values = neighbor_fe_values.get();
+    return *neighbor_fe_values;
+  }
+
+
+
+  template <int dim, int spacedim>
+  const FEFaceValues<dim, spacedim> &
+  ScratchData<dim, spacedim>::reinit_neighbor(
+    const typename DoFHandler<dim, spacedim>::active_cell_iterator &cell,
+    const unsigned int                                              face_no)
+  {
+    if (!neighbor_fe_face_values)
+      neighbor_fe_face_values = std::make_unique<FEFaceValues<dim, spacedim>>(
+        *mapping, *fe, face_quad, neighbor_face_flags);
+    neighbor_fe_face_values->reinit(cell, face_no);
+    cell->get_dof_indices(neighbor_dof_indices);
+    current_neighbor_fe_values = neighbor_fe_face_values.get();
+    return *neighbor_fe_face_values;
+  }
+
+
+
+  template <int dim, int spacedim>
+  const FEFaceValuesBase<dim, spacedim> &
+  ScratchData<dim, spacedim>::reinit_neighbor(
+    const typename DoFHandler<dim, spacedim>::active_cell_iterator &cell,
+    const unsigned int                                              face_no,
+    const unsigned int                                              subface_no)
+  {
+    if (subface_no != numbers::invalid_unsigned_int)
```

Comment:
I don't particular care for this special treatment. What's the reason why someone would not want to just call `reinit (cell, face_no)` in that case?

Reply:
In this way this is nicely compatible with the `mesh_loop` function, which calls face workers with three arguments by default (cell, face number, and subface number), setting the subface number to numbers::invalid_unsigned_int when it is the full face.

Reply:
Oh, I see. I might have complained about that as well if it had gone through patch review at the time (I suspect you modeled that on `MeshWorker`'s behavior).

Reply:
In other words, if that's an idiom used elsewhere already, then let's not worry about it.